### ISSUE TITLE
pydrive2.auth.GoogleAuth LoadCredentials() and SaveCredentials() backend question

### ISSUE THREAD

**Issue by stock90975** [iterative/PyDrive2#115]
Hi,

I would like to use PyDrive2 on my web app where multiple users may be using my app to read/write to their google drive at the same moment.

I understand how to use `LoadCredentialsFile()` and `SaveCredentialsFile()` but I would like to load/save users credentials **NOT** from a file, but from perhaps a database, since I don't want users to read/write over each others `credentials.json`.

Is [LoadCredentials()](https://iterative.github.io/PyDrive2/docs/build/html/pydrive2.html#pydrive2.auth.GoogleAuth.LoadCredentials) or [SaveCredentials()](https://iterative.github.io/PyDrive2/docs/build/html/pydrive2.html#pydrive2.auth.GoogleAuth.SaveCredentials) the solution? If so, what does the `backend` parameter mean?

Otherwise what are possible solutions?

Thanks!

**Comment by shcheklein** [comment]
I think only `file` is supported as a backend now.

**Comment by stock90975** [comment]
Ok, thanks for clearning that up.

I'm super confused with the OAuth flow ... Just to clarify, each unique user that has authorzed my app will need his own `credentials.json`, right?

**Comment by shcheklein** [comment]
>  Just to clarify, each unique user that has authorzed my app will need his own credentials.json, right?

yes and this is not the way you would want to do an actual back end app. Ideally those credentials should be saved into DB, properly encrypted, etc.

I would say PyDrive is quite suboptimal for actual use on the backend the way it's implemented now.

Unless I'm missing something/

**Comment by stock90975** [comment]
Thanks Shcheklein,

I'm thinking of writing the credentials to temp files using NamedTemporaryFile, then grabbing the credentials and putting them properly in a database.

Too bad that LoadCredentials() & SavedCredentials() can't be passed a json string or something like that. Would that be a good feature for PyDrive2?

**Comment by shcheklein** [comment]
> then grabbing the credentials and putting them properly in a database.

yep, but you will have to read them as well. Which can be annoying with temp files, etc. You would need to manage those tmp files carefully for security reasons. Etc.

> Too bad that LoadCredentials() & SavedCredentials() can't be passed a json string or something like that. Would that be a good feature for PyDrive2?

it can be a good feature yes. I'm not sure from the top of my head how specifically it should be implemented - a separate backend ("string"?) or just change those functions? You can do a draft PR if you want and I'll review it to brainstorm together.

**Comment by stock90975** [comment]
So digging into further into
pydrive2/auth.py
https://github.com/iterative/PyDrive2/blob/4328ef3319f282c3aa1baf1333fbed632eab573e/pydrive2/auth.py#L348
```
...
from oauth2client.file import Storage
...
def LoadCredentialsFile(self, backend=None):
    ...
    storage = Storage(credentials_file)
    self.credentials = storage.get()
```
I can't find the exact function in:
https://github.com/googleapis/oauth2client/blob/master/oauth2client/file.py
if this is even the correct module ... because it doesn't have a get() method,  but I'm going to assume it is similar, we've got .locked_get():
https://github.com/googleapis/oauth2client/blob/50d20532a748f18e53f7d24ccbe6647132c979a9/oauth2client/file.py#L35
```
class Storage(client.Storage):
...
def locked_get(self):
        """Retrieve Credential from file. ..."""
        ...
        f = open(self._filename, 'rb')
```
So it appears to me that pydrive2.auth.LoadCredentialsFile() probably relies on oauth2client.file.Storage.get(), or at least some version of .get() that _really, actually_ reads from a file.

Going futher down the rabbit hole ... we've got [oauth2client deprecation!](https://google-auth.readthedocs.io/en/latest/oauth2client-deprecation.html)
> The long-term replacement for oauth2client is this library, google-auth.

> Because we reduced the scope of the library, there are several features in oauth2client we intentionally are not supporting in the v1.0.0 release of google-auth.

> The unsupported features are:
> Support for obtaining user credentials.
> Support for storing credentials.

So unless I'm not understanding some detail, I think implementing my suggestion would at best be a dodgy fix as:
- security wise, one way or another, the credentials will end up in a file, even if momentarily. We can only try to work around this as a possible solution.
- oauth2client is being deprecated, which is where the problem _really_ lies. ie. the "they did it first" defence that we use on our siblings :) ... and PyDrive2 moving away from oauth2client ???!!!
